articles/intraprocess_communication.md

```diff
@@ -0,0 +1,494 @@
+This is particularly true for the default RMW implementation, Fast-RTPS, where the memory requirement increases almost expontentially with the number of participants and entities.
+
+#### Latency and CPU utilization
+
+Publishing a meta-message has the same overhead as that of publishing a small inter-process message.
+
+However, comparing the publication/reception of an intra and an inter-process message, the former requires several additional operations: it has to store the message in the ring buffer, monitor the number of `Subscription`s, and extract the message.
+
+The result is that from the latency and CPU utilization point of view, it is convenient to use intra-process communication only when the message size is at least 5KB.
+
+
+### Problems when both inter and intra-process communication are needed
+
+Currently, ROS2 does not provide any API for making nodes or `Publisher` and `Subscription` to ignore each other.
+This feature would be useful when both inter and intra-process communication are needed.
+
+The reason is that the current implementation of the ROS2 middleware will try to deliver inter-process messages also to the nodes within the same process of the `Publisher`, even if they should have received an intra-process message.
+Note that these messages will be discarded, but they will still cause an overhead.
+
+The DDS specification provides ways for potentially fixing this problem, i.e. with the `ignore_participant`, `ignore_publication` and `ignore_subscription`operations.
+Each of these can be used to ignore a remote participant or entity, allowing to behave as that remote participant did not exist.
+
+The current intra-process communication uses meta-messages that are sent through the RMW between nodes in the same process.
+This has two consequences: first it does not allow to directly "ignore" participants in the same process, because they still have to communicate in order to send and receive meta-messages, thus requiring a more fine-grained control ignoring specific `Publisher`s and `Subscription`s.
+
+Moreover, the meta-messages could be delivered also to nodes in different processes if they have intra-process communication enabled.
+As before, the messages would be discarded immediately after being received, but they would still affect the performances.
+The overhead caused by the additional publication of meta-messages can be potentially reduced by appending to the intra-process topic names a process specific identifier.
+
+## Proposed implementation
+
+### Overview
+
+The new proposal for intra-process communication addresses all the issues previously mentioned.
+It has been designed with performance in mind, so it avoids any communication through the middleware between nodes in the same process.
+Moreover, it supports all the ROS2 Quality of Service configurations.
```

Comment:
Nope.
Honestly I barely know the specification of these new QoS

Reply:
I also thought about this and I think that lifetime related QoS are the only ones that are much harder to implement in this way than sending a meta-message.

See for example:
- https://community.rti.com/static/documentation/connext-dds/5.2.0/doc/manuals/connext_dds/html_files/RTI_ConnextDDS_CoreLibraries_UsersManual/Content/UsersManual/DEADLINE_QosPolicy.htm
- https://community.rti.com/static/documentation/connext-dds/5.2.3/doc/manuals/connext_dds/html_files/RTI_ConnextDDS_CoreLibraries_UsersManual/Content/UsersManual/LIFESPAN_QoS_Policy.htm

Reply:
I'm thinking that is probably worth restricting intra-process communication to a simple, scenario. This also will allow to better optimize it.

No matter which intra-process implementation we will end up having: as long as it is in the `rclcpp` layer it will require a not negligible amount of work to implement and support quality of service and all this will be at the expenses of optimization which should be the main purpose for enabling this feature.

IMO we should only care about the default DDS settings and whether in such a configuration the behaviors of inter and intra-process are coincident.

If an user wants to specify QoS or DDS features, he should go through the RMW and rely on the optimization implemented in the underlying middleware.

Some DDS vendors already have (CycloneDDS) or are implementing (FastRTPS) shared memory communication.
The intra-process communication at the `rclcpp` layer would also act as a benchmark for the performances of all the DDS, allowing to easily evaluate the overhead introduced by a specific middleware.

Intra-process communication would be a feature that has to be manually enabled by the user (i.e. it will not be a default).

Reply:
Yes, I completely agree that we shouldn't focus on supporting all the QoS policies.
Doing that at `rclcpp` level would be hard.

At some point in the future, we should just wrap an intra-process (or inter-process) API provided by the middle-ware itself (there is a proposal about that, see https://github.com/ros2/design/pull/256).

Reply:
> No matter which intra-process implementation we will end up having: as long as it is in the rclcpp layer it will require a not negligible amount of work to implement and support quality of service and all this will be at the expenses of optimization which should be the main purpose for enabling this feature.

I've said this before, but it would be non-negligible work in both cases, but using the meta message would be significantly simpler for many cases, including these new QoS (in my opinion).

That being said, I think the benefits of the new system are worth it even if we cannot support as many QoS settings as easily.

> Intra-process communication would be a feature that has to be manually enabled by the user (i.e. it will not be a default).

That's not clear to me right now. I think there is a good case for intra-process being on by default.

> At some point in the future, we should just wrap an intra-process (or inter-process) API provided by the middle-ware itself (there is a proposal about that, see #256).

I'm not sure this ever 100% replace intra-process in rclcpp, so until that happens, it's not a great reason to ignore this issue.

Reply:
> I can update the sentence specifying which exactly QoS are supported.

Please do update it, if you can, so we can get this merged. In general, I think it would be good to avoid superlatives or all inclusive statements like this. Instead, enumerating what you support is more likely to age well as things change around the document.